### PR TITLE
Fix EVP_PKEY_asn1_copy

### PR DESCRIPTION
**Pull request by HangzeGao**
Add the copy of the omitted ASN1 public key method

Fixes [#21115](https://github.com/openssl/openssl/issues/21115)

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
This is borderline on what we accept with CLA: trivial

@paulidale ?

**Comment by paulidale** [comment]
I'm fine with trivial.  There isn't another way to do this.

**Comment by tmshort** [comment]
I don't see `sig_print` being copied?

> There isn't another way to do this.

One could play with structure offsets and `memcpy()`; or by saving the `dst` real data (5 items), then `memcpy(dst, src, sizeof(*src))`, then restoring the `dst` data. An optimization would be to have the function and non-function data in different structs that could be copied separately.

Unless the compiler is smart enough to know the out-of-order listing (e.g. `old_priv_decode`) is contiguous memory, but since `sig_print()` is also missing, it can't.

**Comment by slontis** [comment]
In master we are doing:
````
 int  x = dst->x;
 ....
 *dst = *src
 
 dst->x = x;
 ....
 ````
 
Doing the change listed here is probably not trivial..

**Comment by tmshort** [comment]
> Doing the change listed here is probably not trivial..
For what I suggest, agreed.

File: crypto/asn1/ameth_lib.c [hunk 1]
```diff
@@ -264,6 +264,7 @@ void EVP_PKEY_asn1_copy(EVP_PKEY_ASN1_METHOD *dst,
 
     dst->pkey_size = src->pkey_size;
     dst->pkey_bits = src->pkey_bits;
+    dst->pkey_security_bits = src->pkey_security_bits;
 
     dst->param_decode = src->param_decode;
     dst->param_encode = src->param_encode;
```
Comment (bernd-edlinger) [inline crypto/asn1/ameth_lib.c:270]: please add `dst->sig_print = src->sig_print;` after
the line copying `param_print` below

**Review by bernd-edlinger — CHANGES_REQUESTED** [review]

**Comment by tom-cosgrove-arm** [comment]
@HangzeGao Can you update with @bernd-edlinger's suggestion, please?

@openssl/otc So if this is not acceptable under `CLA: trivial`, have we asked for @HangzeGao to [sign and submit an ICLA](https://www.openssl.org/policies/cla.html)?

**Comment by mattcaswell** [comment]
> https://github.com/orgs/openssl/teams/otc So if this is not acceptable under CLA: trivial, have we asked for @HangzeGao to [sign and submit an ICLA](https://www.openssl.org/policies/cla.html)?

We have reviewers saying it isn't trivial, so therefore a CLA is required. @HangzeGao - please can you submit one?

**Comment by HangzeGao** [comment]
>

File: crypto/asn1/ameth_lib.c (hunk 1 above)
Comment (HangzeGao) [reply crypto/asn1/ameth_lib.c:270]: It has been added.

**Review by bernd-edlinger — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

I'd consider it trivial, speaking frankly.

**Comment by HangzeGao** [comment]
> > https://github.com/orgs/openssl/teams/otc So if this is not acceptable under CLA: trivial, have we asked for @HangzeGao to [sign and submit an ICLA](https://www.openssl.org/policies/cla.html)?
> 
> We have reviewers saying it isn't trivial, so therefore a CLA is required. @HangzeGao - please can you submit one?

I have submitted the ICLA. I also closed and reopened it. But the CLA check still failed.

**Comment by mattcaswell** [comment]
> But the CLA check still failed.

It's failing because the email address in the author field of the commit does not match the email address provided in the CLA.

**Comment by bernd-edlinger** [comment]
@HangzeGao please change the Author in your commit to your correct e-mail address like this:
```
git commit --amend --author "test <test@test.com>"
git push -f
```

**Comment by t8m** [comment]
Merged to 1.1.1 branch. Thank you for your contribution.

**Pull request closed without merging**